go/mcap/new_reader.go

```diff
@@ -0,0 +1,510 @@
+		return nil, fmt.Errorf("must supply OnMetadata if MetadataFilter is supplied")
+	}
+	return &CallbackReader{opts: opts}, nil
+}
+
+func (m *ReadOptions) upperTimeBound() uint64 {
+	if m.EndTime == 0 {
+		return math.MaxUint64
+	}
+	return m.EndTime
+}
+
+//
+func (cbr *CallbackReader) Read(reader io.Reader) error {
+	if readSeeker, ok := reader.(io.ReadSeeker); ok {
+		return cbr.readIndexed(readSeeker)
+	}
+	return cbr.readUnindexed(reader)
+}
+
+func (cbr *CallbackReader) readUnindexed(r io.Reader) error {
+	lexer, err := NewLexer(r)
+	if err != nil {
+		return err
+	}
+	schemas := make(map[uint16]*Schema)
+	channels := make(map[uint16]*Channel)
+
+	recordBuf := make([]byte, 1024)
+	for {
+		tokenType, record, err := lexer.Next(recordBuf)
+		if err != nil {
+			return fmt.Errorf("failed to read mcap: %w", err)
+		}
+		if len(record) > len(recordBuf) {
+			recordBuf = record
+		}
+		switch tokenType {
+		case TokenSchema:
+			if cbr.opts.OnMessage == nil {
+				continue
+			}
+			schema, err := ParseSchema(record)
+			if err != nil {
+				return fmt.Errorf("failed to parse schema record: %w", err)
+			}
+			if _, ok := schemas[schema.ID]; !ok {
+				schemas[schema.ID] = schema
+			}
+		case TokenChannel:
+			if cbr.opts.OnMessage == nil {
+				continue
+			}
+			channel, err := ParseChannel(record)
+			if err != nil {
+				return fmt.Errorf("failed to parse channel record: %w", err)
+			}
+			if _, ok := channels[channel.ID]; !ok {
+				schema, ok := schemas[channel.SchemaID]
+				if !ok {
+					return ErrUnknownSchema
+				}
+				if (cbr.opts.ChannelFilter != nil) && cbr.opts.ChannelFilter(schema, channel) {
+					channels[channel.ID] = channel
+				}
+			}
+		case TokenMessage:
+			if cbr.opts.OnMessage == nil {
+				continue
+			}
+			message, err := ParseMessage(record)
+			if err != nil {
+				return fmt.Errorf("failed to parse message record: %w", err)
+			}
+			if _, ok := channels[message.ChannelID]; !ok {
+				// skip messages on channels we don't know about. Note that if
+				// an unindexed reader encounters a message it would be
+				// interested in, but has not yet encountered the corresponding
+				// channel ID, it has no option but to skip.
+				continue
+			}
+			if message.LogTime <= cbr.opts.StartTime || message.LogTime > cbr.opts.upperTimeBound() {
+				// timestamp out of bounds
+				continue
+			}
+			channel := channels[message.ChannelID]
+			schema := schemas[channel.SchemaID]
+			if err := cbr.opts.OnMessage(schema, channel, message); err != nil {
+				return err
+			}
+		case TokenAttachment:
+			if cbr.opts.OnAttachment == nil {
+				continue
+			}
+			attachment, err := ParseAttachment(record)
+			if err != nil {
+				return fmt.Errorf("failed to parse attachment record: %w", err)
+			}
+			if attachment.LogTime <= cbr.opts.StartTime || attachment.LogTime > cbr.opts.upperTimeBound() {
+				continue
+			}
+
+			if err := cbr.opts.OnAttachment(attachment); err != nil {
```

Comment:
I think the call to OnAttachment and probably OnMetadata too, needs to get pushed all the way down to the lexer and actually act against the underlying Reader (though the callback can still accept an Attachment type for ergonomics, if the Attachment type is changed to expose a reader instead of a byte slice).

The reason is, the point of the callback-based flow for handling attachments is to avoid materializing them in RAM if not needed, and by the time we are getting tokens out of the lexer the records have already been materialized. If you imagine someone has attached a 20GB file to the mcap, the lexer would probably consumed all the client's ram by materializing token.

I think this matters less/not at all with Message/Statistics.

Reply:
the behavior of the lexer would then be to stream copy the record to io.Discard if the callback is nil. I think that's what happens today for all attachments.

Reply:
Agreed on attachment. Really wrt. metadata? how much metadata are we expecting?

Reply:
Also - i want to do this in a separate PR, i can't figure out how to do that in this PR without totally rewriting parsing, or leaving it in a gross state.

Reply:
I agree metadata is also debatable, I think attachment is the one that definitely will require it.

Reply:
regarding second PR - I think it's fine for it to be in a separate PR, but I'm not sure if this one really stands alone without the change. I think the goal for this change is primarily to support handling of attachments in a streaming manner, with the callback API (for attachments at least) being an anticipated side effect of that, rather than the callback API being a worthy goal in itself. I am assuming in saying that that the end user is still going to end up with something that looks pretty similar to the iterator today, but perhaps implemented on top of these callbacks (btw - how exactly does this work? callbacks that pass messages over channels?)

If we merge the change to cut the reader over to all callbacks prior to streaming attachment support going in, then it seems like we're exposing a version of attachment support that won't work well on large attachments, which I'm not sure is better than just waiting for the streaming support.

I'm not sure about the parsing rewrite, or what counts as gross exactly.. but I think if we wanted to make the change for streaming in the most minimal way, the model to follow would be the way chunks get lexed here: https://github.com/foxglove/mcap/blob/main/go/mcap/lexer.go#L141,L147.

Like attachments, chunk lexing also needs to work against the reader directly unless the EmitChunks option gets specified. So we are checking for a chunk opcode prior to actually reading the body of whatever record we have in hand. Perhaps we need to expand that check for OpChunk into a switch statement that calls l.options.OnAttachment on any attachment record, the similarly falls through to `continue` to actually return the next emitted record from Next. The Attachment case for the lower switch statement would also be removed.

OnAttachment would be passed in as a new element of LexerOptions, with the default value being a callback that writes the record to io.Discard. It would probably get to the lexer through a new options mechanism for readers. https://golang.cafe/blog/golang-functional-options-pattern.html would be a nice pattern to follow for that - some day it would be nice to move the lexer options implementation over to that too.

I think the only piece of parsing this would break is ParseAttachment - which I think would probably either get removed completely or would only be called as an internal piece of the callback mechanism.

That would give us attachment support in unindexed reads, which I think is a good incremental change. It would still leave out,
* Support for attachment callbacks in indexed reads (using the attachment index)
* Support for metadata

Those pieces could be combined into the change or separately addressed; no particularly strong feelings on that.

Anyway, that's just the rough picture I had in my head. Not sure where it breaks down/differs from what you have in mind.